test/support/generators.ex

```diff
@@ -0,0 +1,98 @@
+defmodule SchematicTest.Generators do
+  @moduledoc """
+  Generators for property-based testing
+  """
+
+  import Schematic
+
+  def json_primitive() do
+    StreamData.one_of([
+      StreamData.binary(),
+      StreamData.integer(),
+      StreamData.float(),
+      StreamData.boolean(),
+      StreamData.constant(nil)
+    ])
+  end
+
+  def json_value() do
+    StreamData.one_of([
+      json_primitive(),
+      StreamData.map_of(StreamData.binary(), json_primitive()),
+      StreamData.list_of(json_primitive())
+    ])
+  end
+
+  # TODO: find a non-heinous way to generate tuple schematics
+  @doc """
+  Generator for random `Schematic`s.
+
+  ## Options
+
+  * `:excluding` - a list of `Schematic.kind` values that should **not** be generated. Defaults to `[]`.
+     * e.g. `["null", "int"]`
+  """
+  def schematic(opts \\ []) do
+    excluding = Keyword.get(opts, :excluding, [])
+
+    [
+      int(),
+      bool(),
+      str(),
+      list(),
+      null(),
+      map(
+        key_schematic: str(),
+        value_schematic:
+          StreamData.member_of([
+            str(),
+            int(),
+            bool(),
+            null(),
+            list(),
+            map()
+          ])
+      )
+    ]
+    |> Enum.filter(fn %Schematic{kind: kind} -> kind not in excluding end)
+    |> Enum.map(&StreamData.constant/1)
+    |> StreamData.one_of()
+  end
+
+  def from_schematic(%Schematic{kind: kind}) do
```

Comment:
I think this is closer to what we want.

I think there should probably be some function like `gen` or `generate` that takes a schematic, and then spits out the stream data equivalent.

I think you still need to move the stream data generator for the blueprintable map schematic into this.

Then, I feel like you would be able to to more or less have a single property that generates an arbitrary schematic that can be any depth and combination of them, and test that it unifies and dumps back to itself.

Reply:
And this functionality can probably be re-used to make the user facing stream data API

Reply:
i just added something that is getting towards this idea. it's a little rough, and doesn't cover key transformation for maps, but with this is place we can probably drop a lot of the smaller `property` tests that i wrote along the way to this. i'll take a look at that tomorrow